### DISCUSSION TITLE
Spacy 3.0 - nlp.add_pipe(ruler) failes

### DISCUSSION THREAD

**Discussion by AlexSchmidke** [explosion/spaCy#6265] [Help: Coding & Implementations] [answered]
Hi,

we are trying to use the EntityRuler with the new spacy nightly version. The Code word using the old spacy version but fails at the nlp.add_pipe(ruler) step using spacy nightly 3.0.

The Code (just the example from the spacy website):

```python
import spacy
from spacy.pipeline import EntityRuler

nlp = spacy.load("en_core_web_trf")
ruler = EntityRuler(nlp)
ruler.add_patterns([{"label": "ORG", "pattern": "Apple"}])
nlp.add_pipe(ruler)
```

The same Code runs with the Old Spacy Version (Using the Old Language Set) but fails with following error using spacy nightly

ValueError: [E966] `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component. Expected string, but got <spacy.pipeline.entityruler.EntityRuler object at 0x7f6237e88640> (name: 'None').

- If you created your component with `nlp.create_pipe('name')`: remove nlp.create_pipe and call `nlp.add_pipe('name')` instead.

- If you passed in a component like `TextCategorizer()`: call `nlp.add_pipe` with the string name instead, e.g. `nlp.add_pipe('textcat')`.

- If you're using a custom component: Add the decorator `@Language.component` (for function components) or `@Language.factory` (for class components / factories) to your custom component and assign it a name, e.g. `@Language.component('your_name')`. You can then run `nlp.add_pipe('your_name')` to add it to the pipeline.

## My Environment
copy-paste the result here.-->
* Operating System: Ubuntu 18.04
* Python Version Used: 3.7
* spaCy Version Used: 3.0
* Environment Information:
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
argon2-cffi               20.1.0           py37h7b6447c_1  
async_generator           1.10             py37h28b3542_0  
attrs                     20.2.0                     py_0  
backcall                  0.2.0                      py_0  
bleach                    3.2.1                      py_0  
blis                      0.7.1                    pypi_0    pypi
boto3                     1.15.18                  pypi_0    pypi
botocore                  1.18.18                  pypi_0    pypi
brotlipy                  0.7.0           py37h7b6447c_1000  
ca-certificates           2020.10.14                    0  
catalogue                 2.0.1                    pypi_0    pypi
certifi                   2020.6.20                py37_0  
cffi                      1.14.3           py37he30daa8_0  
chardet                   3.0.4                 py37_1003  
click                     7.1.2                    pypi_0    pypi
cryptography              3.1.1            py37h1ba5d50_0  
cymem                     2.0.3                    pypi_0    pypi
de-core-news-sm           3.0.0a0                  pypi_0    pypi
decorator                 4.4.2                      py_0  
defusedxml                0.6.0                      py_0  
entrypoints               0.3                      py37_0  
idna                      2.10                       py_0  
importlib-metadata        2.0.0                      py_1  
importlib_metadata        2.0.0                         1  
ipykernel                 5.3.4            py37h5ca1d4c_0  
ipython                   7.18.1           py37h5ca1d4c_0  
ipython_genutils          0.2.0                    py37_0  
jedi                      0.17.2                   py37_0  
jinja2                    2.11.2                     py_0  
jmespath                  0.10.0                   pypi_0    pypi
json5                     0.9.5                      py_0  
jsonschema                3.2.0                      py_2  
jupyter_client            6.1.7                      py_0  
jupyter_core              4.6.3                    py37_0  
jupyterlab                2.2.6                      py_0  
jupyterlab_pygments       0.1.2                      py_0  
jupyterlab_server         1.2.0                      py_0  
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20191231         h14c3975_1  
libffi                    3.3                  he6710b0_2  
libgcc-ng                 9.1.0                hdf63c60_0  
libsodium                 1.0.18               h7b6447c_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
markupsafe                1.1.1            py37h14c3975_1  
mistune                   0.8.4           py37h14c3975_1001  
murmurhash                1.0.2                    pypi_0    pypi
nbclient                  0.5.1                      py_0  
nbconvert                 6.0.7                    py37_0  
nbformat                  5.0.8                      py_0  
ncurses                   6.2                  he6710b0_1  
nest-asyncio              1.4.1                      py_0  
notebook                  6.1.4                    py37_0  
numpy                     1.19.2                   pypi_0    pypi
openssl                   1.1.1h               h7b6447c_0  
packaging                 20.4                       py_0  
pandoc                    2.11                 hb0f4dca_0  
pandocfilters             1.4.2                    py37_1  
parso                     0.7.0                      py_0  
pathy                     0.3.1                    pypi_0    pypi
pexpect                   4.8.0                    py37_1  
pickleshare               0.7.5                 py37_1001  
pip                       20.2.3                   py37_0  
preshed                   3.0.2                    pypi_0    pypi
prometheus_client         0.8.0                      py_0  
prompt-toolkit            3.0.8                      py_0  
ptyprocess                0.6.0                    py37_0  
pycparser                 2.20                       py_2  
pydantic                  1.6.1                    pypi_0    pypi
pygments                  2.7.1                      py_0  
pyopenssl                 19.1.0                     py_1  
pyparsing                 2.4.7                      py_0  
pyrsistent                0.17.3           py37h7b6447c_0  
pysocks                   1.7.1                    py37_1  
python                    3.7.9                h7579374_0  
python-dateutil           2.8.1                      py_0  
pytokenizations           0.6.0                    pypi_0    pypi
pyzmq                     19.0.2           py37he6710b0_1  
readline                  8.0                  h7b6447c_0  
requests                  2.24.0                     py_0  
s3transfer                0.3.3                    pypi_0    pypi
send2trash                1.5.0                    py37_0  
setuptools                50.3.0           py37hb0f4dca_1  
six                       1.15.0                     py_0  
smart-open                2.2.1                    pypi_0    pypi
spacy-nightly             3.0.0rc1                 pypi_0    pypi
sqlite                    3.33.0               h62c20be_0  
srsly                     2.3.0                    pypi_0    pypi
terminado                 0.9.1                    py37_0  
testpath                  0.4.4                      py_0  
thinc                     8.0.0rc0                 pypi_0    pypi
tk                        8.6.10               hbc83047_0  
tornado                   6.0.4            py37h7b6447c_1  
tqdm                      4.50.2                   pypi_0    pypi
traitlets                 5.0.5                      py_0  
typer                     0.3.2                    pypi_0    pypi
typing-extensions         3.7.4.3                  pypi_0    pypi
urllib3                   1.25.10                    py_0  
wasabi                    0.8.0                    pypi_0    pypi
wcwidth                   0.2.5                      py_0  
webencodings              0.5.1                    py37_1  
wheel                     0.35.1                     py_0  
xz                        5.2.5                h7b6447c_0  
zeromq                    4.3.3                he6710b0_3  
zipp                      3.3.0                      py_0  
zlib                      1.2.11               h7b6447c_3

**Comment by honnibal** [comment, marked as answer]
Yes this is one of the workflows that's changed in v3. We've tried to make the error message as helpful as possible. If you have a look at the error you pasted, is it clear to you what needs to be done? I thought the wording there was quite good, but I understand it's always hard to guess how things come across to users. You can also find this covered in the migration guide: https://nightly.spacy.io/usage/v3#migrating-add-pipe

**Comment by adrianeboyd** [comment]
Maybe we could provide the string name in the error message for the built-in components in case two? I think a lot of the names aren't obvious from the class names...

**Comment by Giyn** [comment]
I also encountered the same problem when using spacy 3.0. Have you found an effective solution now?

**Reply by adrianeboyd** [reply]
The solution is in the migration guide, however the "string name" part still could be made clearer for users. It would look like this instead:

```python
ruler = nlp.add_pipe("entity_ruler")
ruler.add_patterns(...)
```

See the examples in the sidebar here: https://spacy.io/api/entityruler

**Reply by danilo-css** [reply]
@adrianeboyd you're a life saver. Thanks. Adding to your contribution, I just wanna say this worked for me:

```python
patterns = [
    {"label": "MONEY", "pattern": [{"LOWER": "us$"}, {"LIKE_NUM": True}]},
]

# Add patterns to ruler
ruler.add_patterns(patterns)

# Add the entity ruler to the pipeline
nlp.add_pipe('entity_ruler')
nlp.get_pipe('entity_ruler').add_patterns(patterns)
```

**Comment by rohan2734** [comment]
It still doesnt work for me, It shows the same error 
and Entity ruler also doesnt help me

```
#add a sengmentation rule

# @Language.component('set_custom_boundaries')
def set_custom_boundaries(doc):
    doc.is_parsed = True
#     for token in doc:
    for token in doc[:-1]:
#         print(token)
#         print(token.i)
        if token.text == ';':
            doc[token.i+1].is_sent_start = True
    return doc
        
# set_custom_boundaires(doc)
```
```
# #
# @Language.component("set_custom_boundaries")

nlp.add_pipe(set_custom_boundaries,before='parser')

nlp.pipe_names
```

**Reply by adrianeboyd** [reply]
You need to register the component with the `@Language.component` decorator (not commented out!) and then use the **string** name you used in the decorator with `nlp.add_pipe`, so `nlp.add_pipe("set_custom_boundaries")`.

**Reply by rohan2734** [reply]
I have even tried that too sir, 
this is the code
```
@Language.component("set_custom_boundaries")

nlp.add_pipe("set_custom_boundaries",before='parser')

nlp.pipe_names
```
this is the error
```
  F File "<ipython-input-2-d455df790d88>", line 4
    nlp.add_pipe("set_custom_boundaries",before='parser')
    ^
SyntaxError: invalid syntax
```

I even added the doc.parser=True, and i tried it too,but it is not working.

**Reply by adrianeboyd** [reply]
Please take a look at the docs and examples here for examples of how to use the component decorator: https://spacy.io/usage/processing-pipelines#custom-components

**Reply by rohan2734** [reply]
Hey thankyou it is working now

**Reply by leighton155** [reply]
What about LanguageDetector?  https://spacy.io/universe/project/spacy-langdetect

The online documentation still says to use nlp.add_pipe(LanguageDetector(), name='language_detector', last=True).  Then the error shows `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component.

If I try nlp.add_pipe('langdetect') the error says

ValueError: [E002] Can't find factory for 'langdetect' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, textcat_multilabel, es.lemmatizer, en.lemmatizer

It doesn't appear that Language Detector is an available factory.

**Reply by adidonatocda** [reply]
Not sure if you are still having this issue in Spacy 3.x , I ran into a similar issue and fixed it by adding the following imports:

```
from spacy.language import Language

@Language.component("my_component")
def my_component(doc):
   # Do something to the doc here
   return doc
```

This is outlined in the online docs: https://spacy.io/usage/processing-pipelines#custom-components

**Comment by tushitdave** [comment]
This is also one of the errors when you bring language factory my component. Still thinking how to resolve this but no valid sources so far
TypeError: module, class, method, function, traceback, frame, or code object was expected, got cython_function_or_method

**Reply by polm** [reply]
Could you give a code sample and your full error output? It also might be best for you to open a new discussion for this.

**Reply by tushitdave** [reply]
Hi Polm,

```
#load the model

if model is not None:
    nlp = spacy.load(model)  
    print("Loaded model '%s'" % model)
else:
    nlp = spacy.blank('en')  
    print("Created blank 'en' model")

#set up the pipeline
# @Language.component("ner")
# def ner(train_data):
#     return train_data

@Language.factory("my_component")
def create_my_component(nlp, ner):
    return MyComponent(nlp)

nlp.add_pipe("my_component")
# nlp.add_pipe("ner",last = True)
if 'ner' not in nlp.pipe_names:
    ner = nlp.create_pipe('ner')
    nlp.add_pipe(ner , last = True)
else:
    ner = nlp.get_pipe('ner')
```


I am using spacy 3.0.0 , windows-10, python - 3.6 version

**Reply by tushitdave** [reply]
complete error is as follows:-

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-123-8eae84ba1be4> in <module>
     10 #set up the pipeline
     11 @Language.component("ner")
---> 12 def ner(train_data):
     13     return train_data
     14 

~\AppData\Roaming\Python\Python38\site-packages\spacy\language.py in add_component(component_func)
    553                     factory_func = existing_func  # noqa: F811
    554 
--> 555             cls.factory(
    556                 component_name,
    557                 assigns=assigns,

~\AppData\Roaming\Python\Python38\site-packages\spacy\language.py in factory(cls, name, default_config, assigns, requires, retokenizes, default_score_weights, func)
    497 
    498         if func is not None:  # Support non-decorator use cases
--> 499             return add_factory(func)
    500         return add_factory
    501 

~\AppData\Roaming\Python\Python38\site-packages\spacy\language.py in add_factory(factory_func)
    463                 # if module is reloaded.
    464                 existing_func = registry.factories.get(internal_name)
--> 465                 if not util.is_same_func(factory_func, existing_func):
    466                     err = Errors.E004.format(
    467                         name=name, func=existing_func, new_func=factory_func

~\AppData\Roaming\Python\Python38\site-packages\spacy\util.py in is_same_func(func1, func2)
    932         return False
    933     same_name = func1.__qualname__ == func2.__qualname__
--> 934     same_file = inspect.getfile(func1) == inspect.getfile(func2)
    935     same_code = inspect.getsourcelines(func1) == inspect.getsourcelines(func2)
    936     return same_name and same_file and same_code

~\anaconda3\envs\spacy\lib\inspect.py in getfile(object)
    674     if iscode(object):
    675         return object.co_filename
--> 676     raise TypeError('module, class, method, function, traceback, frame, or '
    677                     'code object was expected, got {}'.format(
    678                     type(object).__name__))

TypeError: module, class, method, function, traceback, frame, or code object was expected, got cython_function_or_method
```

**Reply by polm** [reply]
OK, thanks for the details. Please [read the Markdown guide](https://guides.github.com/features/mastering-markdown/) and use the Preview function to make sure your post is readable. 

I'm not sure how you're defining your component but I guess it's a Cython cdef class or possibly function? You need to make it a Python object/function, like the error says. Can you show the code for MyComponent?

Also please upgrade spaCy to the latest instead of 3.0.0 - we've added several bugfixes since then.

**Reply by polm** [reply]
Wait. I'm confused now. Based on the stack trace, the error is happening in code you commented out in your code sample. What code are you actually running?

```
#set up the pipeline
@Language.component("ner")
def ner(train_data):
    return train_data
```

You probably shouldn't name a component "ner" as that might conflict with the existing pipelines. You also have a separate, unrelated `ner` variable later in your code, so that function name should be changed (if you're actually using that code).  Also, since this is a stateless component, the argument for it is a Doc, not training data. See [the docs](https://spacy.io/usage/processing-pipelines#custom-components-simple).

**Reply by tushitdave** [reply]
Hi , This is in reference to the below code, I am unable to create custom pipeline(input and output both are presented in the code).
Input:-
def add_new_ruler(skill_pattern_path):
new_ent_ruler = EntityRuler(nlp).from_disk(skill_pattern_path)
new_ent_ruler = nlp.create_pipe('new_ent_ruler',new_ent_ruler)
nlp.add_pipe(new_ent_ruler, after='parser' , before ='ner')

print(nlp.pipe_names)

output:-
spacy pipeline issue
['tok2vec', 'tagger', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']

python version 3.8.5
spacy version: 3.0.0
windows:10
![spacy pipeline issue](https://user-images.githubusercontent.com/52331739/123927476-82f9aa80-d9aa-11eb-8f24-d9682046153e.PNG)

**Reply by tushitdave** [reply]
@polm - I am able to successfully run the code however I can't see my customized pipeline after the parser and before the NER. Is there anything I am missing here

**Reply by svlandeg** [reply]
I only see this now, but we can continue the discussion in this specific thread you opened, to avoid duplicating stuff: https://github.com/explosion/spaCy/discussions/8550
